### PR TITLE
GraphQL v7.5.0 instrumentation

### PR DESCRIPTION
**Pull request by RassK**
## Why

Fixes https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2464

## What

* Adds native support for GraphQL version 7.5.0+
* Removes existing versions [2.3-2.4] support.

## Tests

* Added new set of integration tests

## Follow-ups

* https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2699

## Checklist

- [x] `CHANGELOG.md` is updated.
- [x] Documentation is updated.
- [x] New features are covered by tests.

### CODE REVIEW

File: test/IntegrationTests/LibraryVersions.g.cs [hunk 1]
```diff
@@ -47,6 +47,14 @@ public static class LibraryVersion
 #else
         new object[] { "2.3.0" },
         new object[] { "2.4.0" },
+#endif
+    };
+    public static readonly IReadOnlyCollection<object[]> GraphQLNativeSupport = new List<object[]>
```
Comment (pjanotti) [inline test/IntegrationTests/LibraryVersions.g.cs:52]:
```suggestion
    public static readonly IReadOnlyCollection<object[]> GraphQLNativeInstrumentation = new List<object[]>
```

Comment (pjanotti) [reply test/IntegrationTests/LibraryVersions.g.cs:52]: or perhaps
```suggestion
    public static readonly IReadOnlyCollection<object[]> GraphQLNativelyInstrumented = new List<object[]>
```

File: test/IntegrationTests/GraphQLByteCodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    {
+    }
+
+    public override bool SupportsExceptions => true;
+
+    public override string InstrumentationScope => "OpenTelemetry.AutoInstrumentation.GraphQL";
+
+    [Theory]
+    [Trait("Category", "EndToEnd")]
+    [MemberData(nameof(GetData), new object[] { false })]
```
Comment (pjanotti) [inline test/IntegrationTests/GraphQLByteCodeTests.cs:34]: The `MemberData` here is pretty opaque, a comment or a constant to convey the meaning of the boolean would be helpful.

Comment (pjanotti) [reply test/IntegrationTests/GraphQLByteCodeTests.cs:34]: Same for the "*.NativeTests.cs"

File: test/IntegrationTests/GraphQLNativeTests.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+// limitations under the License.
+// </copyright>
+
+#if NET6_0_OR_GREATER
+
+using Xunit.Abstractions;
+
+namespace IntegrationTests;
+
+public class GraphQLNativeTests : GraphQLTests
```
Comment (pjanotti) [inline test/IntegrationTests/GraphQLNativeTests.cs:23]: Similar to other suggested renames:
```suggestion
public class GraphQLNativelyInstrumented : GraphQLTests
```

**Review by pjanotti — COMMENTED** [review]
LGTM - a few naming suggestions, and an important Q about `netstandard2.0`.

File: test/test-applications/integrations/TestApplication.GraphQL/TestApplication.GraphQL.csproj [hunk 4]
```diff
@@ -1,10 +1,14 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk.Web">
```
[the remark this answers is not in the document]
Comment (RassK) [reply test/test-applications/integrations/TestApplication.GraphQL/TestApplication.GraphQL.csproj:1]: @pjanotti thinking now to remove all that "natively" and instead rename the old project `TestApplication.GraphQL.Legacy` so the current (newest) way will have "normal" name.

File: test/IntegrationTests/GraphQLByteCodeTests.cs (hunk 2 above)
Comment (RassK) [reply test/IntegrationTests/GraphQLByteCodeTests.cs:34]: @pjanotti I was also concerned about this. Either a new enum or 

```csharp
new object[] { /* isNativelyInstrumented */ false }
```

any preferences?

File: test/test-applications/integrations/TestApplication.GraphQL/TestApplication.GraphQL.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (pjanotti) [reply test/test-applications/integrations/TestApplication.GraphQL/TestApplication.GraphQL.csproj:1]: Sounds good.

**Comment by Kielek** [comment]
What's about fully dropping legacy instrumentation?
We will have huge gap between [3.0, 7.5.0) with non-instrumented versions. It is unlikely that we would like to support these versions.

I do not see any important differences with supported .NET and .NET Frameworl versions for 
https://www.nuget.org/packages/GraphQL/7.5.0#supportedframeworks-body-tab
and https://www.nuget.org/packages/GraphQL/2.4.0#supportedframeworks-body-tab

File: docs/internal/instrumentation-libraries.md
Comment (Kielek) [inline docs/internal/instrumentation-libraries.md:1]: You have forgot also about https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/54804fef53e9522113bd77392fbbd5d80fec015d/docs/config.md#traces-instrumentations

**Comment by RassK** [comment]
> What's about fully dropping legacy instrumentation? We will have huge gap between [3.0, 7.5.0) with non-instrumented versions. It is unlikely that we would like to support these versions.
> 
> I do not see any important differences with supported .NET and .NET Frameworl versions for https://www.nuget.org/packages/GraphQL/7.5.0#supportedframeworks-body-tab and https://www.nuget.org/packages/GraphQL/2.4.0#supportedframeworks-body-tab

@Kielek I think I would be fine dropping [2.0.0, 2.4.0]. Just wondering that 2.4.0 is by far the most popular version and it's around 4.5 years old. Kind of in the border range to consider as legacy. 

@Shane32 you may have better insights whether version range [2.0.0, 2.4.0] is still topical?

image

**Comment by Shane32** [comment]
2.4.0 was (is?) used by some large legacy systems. Occasionally we get asked questions about that version but it is getting rare now. Those users seem to be hesitant to upgrade - again probably because it’s a legacy system, but also because 3.0 changed some fundamentals.

Also, there was a very large time gap (and feature gap) between 2.4 and 3.0 which is why there are so many more downloads of 2.4 than newer versions. Since 3.0 there has been a fairly regular release cadence, so no single version amassed a great quantity of downloads. For instance, 4.x versions total nearly the same number of downloads as 2.x. 

We also do not provide “support” for 2.4 any more, although we usually try to answer questions for newer versions when possible. Generally all fixes go into the latest version although it is fairly easy for us to hotfix 4.x and newer if a critical flaw were identified.  Not so for 2.x

Considering that OpenTelemetry is a fairly new platform (?), I wouldn’t expect users of legacy systems containing 2.4 to be adding OpenTelemetry to it. They likely have legacy telemetry systems. Or, I’d expect them to be trying to keep all their dependencies up to date. That’s just a guess.

**Comment by Kielek** [comment]
@Shane32, this project is able to support only officially supported .NET Framework and .NET versiosn. It means .NET Framewrok 4.6.2+ and .NET 6+.

Older versions are not supported by OpenTelemetry SDK/API which is our base.

**Comment by Kielek** [comment]
SIG meeting decisions: lets drop 2.x support for now. We are just before GA. If needed we can bring it back later.

File: CHANGELOG.md [hunk 5]
```diff
@@ -33,12 +33,16 @@ This release is built on top of [OpenTelemetry .NET](https://github.com/open-tel
   Both supported only on .NET Framework.
 - `WCFCLIENT` and `WCFSERVICE` no longer requires manual modification
   of config files to include `TelemetryEndpointBehaviorExtensionElement`.
+- [GraphQL](https://www.nuget.org/packages/GraphQL/) instrumentation is now
+  supported from version 7.5.0
```
Comment (Kielek) [inline CHANGELOG.md:37]:
```suggestion
- [GraphQL](https://www.nuget.org/packages/GraphQL/) instrumentation is now
  supported from version 7.5.0.
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -33,12 +33,16 @@ This release is built on top of [OpenTelemetry .NET](https://github.com/open-tel
...
 
 ### Deprecated
 
 ### Removed
 
 - Removed `WCF` instrumentation for Core WCF Client working on .NET.
+- Removed [GraphQL](https://www.nuget.org/packages/GraphQL/) instrumentation
+  for versions 2.0.0-2.4.*
```
Comment (Kielek) [inline CHANGELOG.md:45]: 2.0.0-2.2.0 was never supported.

```suggestion
- Removed [GraphQL](https://www.nuget.org/packages/GraphQL/) instrumentation
  for versions 2.3.0-2.4.*.
```

File: docs/config.md [hunk 7]
```diff
@@ -127,7 +127,7 @@ due to lack of stable semantic convention.
...
 | `ENTITYFRAMEWORKCORE` | [Microsoft.EntityFrameworkCore](https://www.nuget.org/packages/) **Not supported on .NET Framework**                                                                                            | ≥6.0.12            | source                  | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
-| `GRAPHQL`             | [GraphQL](https://www.nuget.org/packages/GraphQL)                                                                                                                                               | ≥2.3.0 & < 3.0.0   | bytecode                | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
+| `GRAPHQL`             | [GraphQL](https://www.nuget.org/packages/GraphQL)                                                                                                                                               | ≥7.5.0             | source                  | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
```
Comment (Kielek) [inline docs/config.md:130]:
```suggestion
| `GRAPHQL`             | [GraphQL](https://www.nuget.org/packages/GraphQL) **Not supported on .NET Framework**                                                                                                           | ≥7.5.0             | source                  | [Experimental](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md) |
```

File: src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs [hunk 8]
```diff
@@ -270,9 +270,9 @@ private static void AddLazilyLoadedMetricInstrumentations(LazyInstrumentationLoa
         }
     }
 
-    private static void AddLazilyLoadedTraceInstrumentations(LazyInstrumentationLoader lazyInstrumentationLoader, PluginManager pluginManager, IReadOnlyList<TracerInstrumentation> enabledInstrumentations)
+    private static void AddLazilyLoadedTraceInstrumentations(LazyInstrumentationLoader lazyInstrumentationLoader, PluginManager pluginManager, TracerSettings tracerSettings)
```
Comment (Kielek) [inline src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs:273]: Why this change?

File: test/IntegrationTests/GraphQLTests.cs
Comment (Kielek) [inline test/IntegrationTests/GraphQLTests.cs:1]: As we dropped bytecode instrumentation, there is no need to these cases.
Consider to simplify test to the level of https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/1f77dc1eb43c4fa350203dea9b56e8ee26404b32/test/IntegrationTests/MySqlDataTests.cs

File: test/test-applications/integrations/TestApplication.GraphQL/Properties/launchSettings.json [hunk 9]
```diff
@@ -12,20 +12,9 @@
       "commandName": "Project",
       "launchBrowser": false,
       "environmentVariables": {
-        "ASPNETCORE_ENVIRONMENT": "Development",
```
Comment (Kielek) [inline test/test-applications/integrations/TestApplication.GraphQL/Properties/launchSettings.json:15]: What is the reason to remove easy way to attach OTel Auto?

File: test/Directory.Packages.props [hunk 10]
```diff
@@ -5,9 +5,10 @@
...
     <PackageVersion Include="FluentAssertions" Version="6.11.0" />
     <PackageVersion Include="Google.Protobuf" Version="3.23.3" />
-    <PackageVersion Include="GraphQL" Version="2.4.0" />
-    <PackageVersion Include="GraphQL.Server.Transports.AspNetCore" Version="3.4.0" />
-    <PackageVersion Include="GraphQL.Server.Ui.Playground" Version="3.4.0" />
+    <PackageVersion Include="GraphQL" Version="7.5.0" />
+    <PackageVersion Include="GraphQL.MicrosoftDI" Version="7.5.0" />
+    <PackageVersion Include="GraphQL.Server.Transports.AspNetCore" Version="7.4.0" />
+    <PackageVersion Include="GraphQL.Server.Ui.Playground" Version="7.4.0" />
     <PackageVersion Include="GraphQL.StarWars" Version="1.0.0" />
```
Comment (Kielek) [inline test/Directory.Packages.props:12]: Remove this line. It is no loner used.

**Comment by Kielek** [comment]
Everything under https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/tree/1f77dc1eb43c4fa350203dea9b56e8ee26404b32/src/OpenTelemetry.AutoInstrumentation/Instrumentations/GraphQL should be removed.

File: src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs (hunk 8 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs:273]: same method & `case TracerInstrumentation.GraphQL:`

File: docs/config.md (hunk 7 above)
Comment (RassK) [reply docs/config.md:130]: Is supported but we don't have tests.

File: test/IntegrationTests/GraphQLTests.cs
Comment (RassK) [reply test/IntegrationTests/GraphQLTests.cs:1]: I started to think here that we did have those tests in .NET SDK side in case of official or community adapter library.
But if library itself is ejecting activities, then if we should also verify OTel compatibility - or just leave it for the library author?

File: test/test-applications/integrations/TestApplication.GraphQL/Properties/launchSettings.json (hunk 9 above)
Comment (RassK) [reply test/test-applications/integrations/TestApplication.GraphQL/Properties/launchSettings.json:15]: part of cleanup, outdated values - actually most of our launchsettings are clean, so I just followed the pattern there.

File: test/IntegrationTests/GraphQLTests.cs
Comment (Kielek) [reply test/IntegrationTests/GraphQLTests.cs:1]: For this kind of instrumentation, we have agreed to only check if instrumentation is working for the simplest case.
Any tags verification should be on library author.

If we decide to verify on our side we should expect different spans for 2.0 and probably 2.3/2.4 when the specification will be stabilized.

File: src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs (hunk 8 above)
Comment (Kielek) [reply src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs:273]: I have missed that. Thanks.

File: docs/config.md (hunk 7 above)
Comment (Kielek) [reply docs/config.md:130]: Fixed b663f953e38f070356c9350ef0fa0977f3fe00ee

File: src/OpenTelemetry.AutoInstrumentation/Configurations/EnvironmentConfigurationTracerHelper.cs [hunk 11]
```diff
@@ -52,6 +52,7 @@ internal static class EnvironmentConfigurationTracerHelper
                 TracerInstrumentation.MySqlData => Wrappers.AddMySqlClientInstrumentation(builder, pluginManager, lazyInstrumentationLoader),
                 TracerInstrumentation.StackExchangeRedis => builder.AddSource("OpenTelemetry.Instrumentation.StackExchangeRedis"),
                 TracerInstrumentation.EntityFrameworkCore => Wrappers.AddEntityFrameworkCoreInstrumentation(builder, pluginManager, lazyInstrumentationLoader),
+                TracerInstrumentation.GraphQL => Wrappers.AddGraphQLInstrumentation(builder, pluginManager, lazyInstrumentationLoader, settings),
```
[the remark this answers is not in the document]
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/EnvironmentConfigurationTracerHelper.cs]: I think we need to make proper .NET Framework test application. Microsoft.AspnetCore is also now deprecated, so prob good time separate ASP.NET Core from running on netfx.

File: CHANGELOG.md [hunk 12]
```diff
@@ -33,12 +33,16 @@ This release is built on top of [OpenTelemetry .NET](https://github.com/open-tel
   Both supported only on .NET Framework.
 - `WCFCLIENT` and `WCFSERVICE` no longer requires manual modification
   of config files to include `TelemetryEndpointBehaviorExtensionElement`.
+- [GraphQL](https://www.nuget.org/packages/GraphQL/) instrumentation is now
+  supported from version 7.5.0.
```
Comment (Shane32) [inline CHANGELOG.md:37]: GraphQL.NET 7.5 only supports OpenTelemetry on .NET 5+ because the necessary APIs do not exist within .NET Framework or .NET Standard.

File: docs/internal/instrumentation-libraries.md [hunk 13]
```diff
@@ -59,7 +59,7 @@ submit an issue to request its inclusion.
...
 | [gRPC-dotnet](https://github.com/grpc/grpc-dotnet)                                |                          Partial[^grpc_client_only]                          |                                                                               | Client and service support should be added.                                                                                                            |
-| [GraphQL](https://www.nuget.org/packages/GraphQL/)                                |                     Partial (versions ≥2.3.0 & < 3.0.0)                      |                                                                               | The current instrumentation needs updates to match the semantic conventions.                                                                           |
+| [GraphQL](https://www.nuget.org/packages/GraphQL/)                                |                                     Yes                                      |                                                                               |                                                                                                                                                        |
```
Comment (Shane32) [inline docs/internal/instrumentation-libraries.md:62]: Specify version?  Specify .NET 5+?

File: src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    public GraphQLInitializer(PluginManager pluginManager, TracerSettings tracerSettings)
+    : base("GraphQL")
+    {
+        _pluginManager = pluginManager;
+        _tracerSettings = tracerSettings;
+    }
+
+    public override void Initialize(ILifespanManager lifespanManager)
+    {
+        var initializerType = Type.GetType("OpenTelemetry.AutoInstrumentation.Initializer, GraphQL");
```
Comment (Shane32) [inline src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs:37]: I'm a bit disappointed that this code is specific to GraphQL.  I would have expected this to be generic to any library -- just add an internal class with the appropriate name to the library to support auto instrumentation (when not automatically emitted by the library).  In other words, it seems that if I write another library and add a similar class named OpenTelemetry.AutoInstrumentation.Initializer, it's not going to work unless there's specific support added to OpenTelemetry.

File: src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        if (initializerType == null)
+        {
+            return;
+        }
+
+        var optionsType = Type.GetType("GraphQL.Telemetry.GraphQLTelemetryOptions, GraphQL")!;
+        var optionsInstance = Activator.CreateInstance(optionsType)!;
```
Comment (Shane32) [inline src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs:44]: The the same vein, I would have expected this to by a dynamic mapping of options -- or, to not support options at all, just as if the library emitted tracing events by default.

File: docs/internal/instrumentation-libraries.md (hunk 13 above)
Comment (RassK) [reply docs/internal/instrumentation-libraries.md:62]: There's different doc for that https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2681/files#diff-77b3519370ff48f6caf7c316f0ece966f2f0871c40e9381d451d2c5b1c76b40fR130

But according to https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2681#discussion_r1238544573, the information has to be adjusted.

File: src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs (hunk 14 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs:37]: Unfortunately we cannot scan every loaded library in the startup time, that's going to add up. 

The automated way is to use `OTEL_DOTNET_AUTO_TRACES_ADDITIONAL_SOURCES` https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/docs/config.md#additional-settings. It's expected that every library in the future will use activities by default without custom bootstrapping.

Initializer is already one of many ways to bootstrap activity broadcasting. Definitely the newest way, still not sure how many are going to adapt this pattern.

File: src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs (hunk 15 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/GraphQLInitializer.cs:44]: This should become better when OTel config file is finally speced. Currently introducing new envs is also on hold.
But in the end environment controller (end user) and OTel vendor (via plugins) must have control over options. Some of the options can control sending sensitive data over the wire, which is extremely important to control.

File: test/IntegrationTests/GraphQLTests.cs
Comment (RassK) [reply test/IntegrationTests/GraphQLTests.cs:1]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2699

**Review by Kielek — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configurations/EnvironmentConfigurationTracerHelper.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Shane32) [reply src/OpenTelemetry.AutoInstrumentation/Configurations/EnvironmentConfigurationTracerHelper.cs:55]: Just FYI, Microsoft.AspNetCore 2.1 is supported for .NET Framework. 

https://dotnet.microsoft.com/en-us/platform/support/policy/aspnet/2.1-packages
